src/systems/systems/src/shared/pid.rs

```diff
@@ -0,0 +1,173 @@
+use std::time::Duration;
+
+// Pid controller implementation
+// Implementation in a recursive form
+// u(k+1) = u(k) + (e(k)-e(k-1))*Kp + e(k) * Ki * dt + (e(k)- 2*e(k-1) + e(k-2)) * Kd / dt
+// For variable dt add dt duration as argument, if fixed dt or scheduled synchronously with controlled system input None as dt
+pub struct Pid {
+    kp: f64,
+    ki: f64,
+    kd: f64,
+
+    min_output: f64,
+    max_output: f64,
+
+    setpoint: f64,
+
+    error_k_1: Option<f64>,
+    error_k_2: Option<f64>,
+    output: f64,
+}
+
+impl Pid {
+    pub fn new(kp: f64, ki: f64, kd: f64, min_output: f64, max_output: f64, setpoint: f64) -> Self {
+        Self {
+            kp,
+            ki,
+            kd,
+            min_output,
+            max_output,
+            setpoint,
+            error_k_1: None,
+            error_k_2: None,
+            output: 0.,
+        }
+    }
+
+    pub fn change_setpoint(&mut self, new_setpoint: f64) {
+        self.setpoint = new_setpoint;
+    }
+
+    pub fn setpoint(&self) -> f64 {
+        self.setpoint
+    }
+
+    pub fn reset(&mut self, setpoint: f64) {
+        self.output = 0.;
+        self.change_setpoint(setpoint);
+        self.reset_error();
+    }
+
+    pub fn next_control_output(&mut self, measurement: f64, delta_time: Option<Duration>) -> f64 {
+        let mut dt = 1.;
+        if let Some(delta) = delta_time {
+            dt = delta.as_secs_f64();
+        }
+
+        let error = self.setpoint - measurement;
+
+        let p_term;
+        if self.error_k_1.is_some() {
+            p_term = (error - self.error_k_1.unwrap()) * self.kp;
+        } else {
+            p_term = error * self.kp;
+        }
```

Comment:
```suggestion
        let mut p_term = error * self.kp;
        if let Some(error_k_1) = self.error_k_1 {
            p_term -= error_k_1 * self.kp;
        }
```

Reply:
Wouldn't we add another multiplication here just for fun? I know cpu is cheap :D    Must admit I'm not a fan of the let Some sugar stuff.

Reply:
As you say, one multiplication is really not something we should be concerned about. I can live with the `is_some` -> `unwrap` stuff, but I wouldn't feel good if I hadn't point it out :D